runtime/moonbase/src/precompiles.rs

```diff
@@ -86,48 +86,135 @@ parameter_types! {
 	// PrecompileAt<AddressU64<1025>, Dispatch<R>>,
-	PrecompileAt<AddressU64<1026>, ECRecoverPublicKey>,
+	PrecompileAt<AddressU64<1026>, ECRecoverPublicKey, (CallableByContract, CallableByPrecompile)>,
 	// Moonbeam specific precompiles:
-	PrecompileAt<AddressU64<2048>, ParachainStakingPrecompile<R>>,
-	PrecompileAt<AddressU64<2049>, CrowdloanRewardsPrecompile<R>>,
-	PrecompileAt<AddressU64<2050>, Erc20BalancesPrecompile<R, NativeErc20Metadata>>,
-	PrecompileAt<AddressU64<2051>, DemocracyPrecompile<R>>,
-	PrecompileAt<AddressU64<2052>, XtokensPrecompile<R>>,
-	PrecompileAt<AddressU64<2053>, RelayEncoderPrecompile<R, WestendEncoder>>,
-	PrecompileAt<AddressU64<2054>, XcmTransactorPrecompileV1<R>>,
-	PrecompileAt<AddressU64<2055>, AuthorMappingPrecompile<R>>,
-	PrecompileAt<AddressU64<2056>, BatchPrecompile<R>, LimitRecursionTo<2>>,
-	PrecompileAt<AddressU64<2057>, RandomnessPrecompile<R>>,
-	PrecompileAt<AddressU64<2058>, CallPermitPrecompile<R>>,
-	PrecompileAt<AddressU64<2059>, ProxyPrecompile<R>>,
-	PrecompileAt<AddressU64<2060>, XcmUtilsPrecompile<R, XcmExecutorConfig>>,
-	PrecompileAt<AddressU64<2061>, XcmTransactorPrecompileV2<R>>,
-	PrecompileAt<AddressU64<2062>, CollectivePrecompile<R, CouncilInstance>>,
-	PrecompileAt<AddressU64<2063>, CollectivePrecompile<R, TechCommitteeInstance>>,
-	PrecompileAt<AddressU64<2064>, CollectivePrecompile<R, TreasuryCouncilInstance>>,
+	PrecompileAt<
+		AddressU64<2048>,
+		ParachainStakingPrecompile<R>,
+		(CallableByContract, CallableByPrecompile),
+	>,
+	PrecompileAt<
+		AddressU64<2049>,
+		CrowdloanRewardsPrecompile<R>,
+		(CallableByContract, CallableByPrecompile),
+	>,
+	PrecompileAt<
+		AddressU64<2050>,
+		Erc20BalancesPrecompile<R, NativeErc20Metadata>,
+		(CallableByContract, CallableByPrecompile),
+	>,
+	PrecompileAt<
+		AddressU64<2051>,
+		DemocracyPrecompile<R>,
+		(CallableByContract, CallableByPrecompile),
+	>,
+	PrecompileAt<
+		AddressU64<2052>,
+		XtokensPrecompile<R>,
+		(CallableByContract, CallableByPrecompile),
+	>,
+	PrecompileAt<
+		AddressU64<2053>,
+		RelayEncoderPrecompile<R, WestendEncoder>,
+		(CallableByContract, CallableByPrecompile),
+	>,
+	PrecompileAt<
+		AddressU64<2054>,
+		XcmTransactorPrecompileV1<R>,
+		(CallableByContract, CallableByPrecompile),
+	>,
+	PrecompileAt<
+		AddressU64<2055>,
+		AuthorMappingPrecompile<R>,
+		(CallableByContract, CallableByPrecompile),
+	>,
+	PrecompileAt<
+		AddressU64<2056>,
+		BatchPrecompile<R>,
+		(
+			SubcallWithMaxNesting<2>,
+			// Batch is the only precompile allowed to call Batch.
+			CallableByPrecompile<OnlyFrom<AddressU64<2056>>>,
+		),
+	>,
+	PrecompileAt<
+		AddressU64<2057>,
+		RandomnessPrecompile<R>,
+		(SubcallWithMaxNesting<0>, CallableByContract),
+	>,
+	PrecompileAt<
+		AddressU64<2058>,
+		CallPermitPrecompile<R>,
+		(SubcallWithMaxNesting<0>, CallableByContract),
+	>,
+	PrecompileAt<
+		AddressU64<2059>,
+		ProxyPrecompile<R>,
+		(CallableByContract<OnlyIsProxy<R>>, SubcallWithMaxNesting<0>),
```

Comment:
```suggestion
		(
		    CallableByContract<OnlyIsProxy<R>>,
		    SubcallWithMaxNesting<0>,
		    // Batch is the only precompile allowed to call Proxy.
		    CallableByPrecompile<OnlyFrom<AddressU64<2056>>>,
		),
```

If we need to execute several calls on behalf of the proxy account in one tx, we need to allow the batch precompile to call the proxy precompile.

Reply:
For that case wouldn't it be better to have a proxy to a batch of many calls? I guess both can be allowed if you want to batch calls through different proxies.

Reply:
For security reasons proxied call are very limited, the list of whitelisted precompiles not contains Batch nor any precompile that can made a subcall

Reply:
It's to not allow to workaround the proxy type filter